ext2/dir.c

```diff
@@ -129,11 +129,16 @@ int _ext2_dir_read(ext2_t *fs, ext2_obj_t *dir, offs_t offs, struct dirent *res,
 	if ((entry = (ext2_dirent_t *)malloc(len)) == NULL)
 		return -ENOMEM;
 
-	if ((ret = _ext2_file_read(fs, dir, offs, (char *)entry, len)) != len) {
+	if ((ret = _ext2_file_read(fs, dir, offs, (char *)entry, len)) < sizeof(ext2_dirent_t)) {
```

Comment:
nice error due to implicit integer promotion :).
- assume `ret = -1` 
-  You're comparing against `size_t` (unsigned type), `ret` will be promoted to unsigned type - UINT_MAX value
- the actual comparison (`UINT_MAX < sizeof(ext_dirent_t)`) would be **false**

See https://wiki.sei.cmu.edu/confluence/display/c/INT02-C.+Understand+integer+conversion+rules for more info and how it should be fixed.

(note: previous code was also invalid, but due to using `!=` would not fail so spectacularly :))

Reply:
I've made a fix. Can you check if it is correct now?